transaction/undoTx.go

```diff
@@ -205,18 +173,53 @@ type sliceHeader struct {
 	cap  int
 }
 
-func (t *undoTx) ReadLog(interface{}) (retVal interface{}) {
-	// undoTx doesn't support this. TODO: Should we panic here?
+func (t *undoTx) ReadLog(intf ...interface{}) (retVal interface{}) {
+	if len(intf) == 2 {
+		return t.readSliceElem(intf[0], intf[1].(int))
+	} else if len(intf) == 3 {
+		return t.readSlice(intf[0], intf[1].(int), intf[2].(int))
+	} else if len(intf) != 1 {
+		panic("[undoTx] ReadLog: Incorrect number of args passed")
+	}
+	ptrV := reflect.ValueOf(intf[0])
+	switch ptrV.Kind() {
+	case reflect.Ptr:
+		data := reflect.Indirect(ptrV)
+		retVal = data.Interface()
+	default:
+		panic("[undoTx] ReadLog: Arg must be pointer")
+	}
 	return retVal
 }
 
-func (t *undoTx) Log(data ...interface{}) error {
-	if len(data) != 1 {
-		return errors.New("[undoTx] Log: Incorrectly used. Correct usage: " +
-			"Log(ptr) OR Log(slice)")
+func (t *undoTx) readSliceElem(slicePtr interface{}, index int) interface{} {
+	ptrV := reflect.ValueOf(slicePtr)
```

Comment:
Should we check if slicePtr is indeed a slice here? Wouldn't we access unsafe memory if user passes a non-slice datastructure as argument 0?

Reply:
Value.Index() method of reflect package used later (on line 198) panics if s's kind is not a slice. So we should catch if slicePtr doesn't point to slice.
https://golang.org/pkg/reflect/#Value.Index